flow/layers/container_layer.cc

```diff
@@ -6,13 +6,42 @@
 
 namespace flutter {
 
-ContainerLayer::ContainerLayer() {}
+ContainerLayer::ContainerLayer()
+    : child_needs_screen_readback_(false), renders_to_save_layer_(false) {}
 
 ContainerLayer::~ContainerLayer() = default;
 
 void ContainerLayer::Add(std::shared_ptr<Layer> layer) {
-  layer->set_parent(this);
+  Layer* the_layer = layer.get();
+  the_layer->set_parent(this);
   layers_.push_back(std::move(layer));
+  if (the_layer->tree_reads_surface()) {
+    UpdateChildReadback(the_layer);
+  }
+}
+
+void ContainerLayer::UpdateChildReadback(Layer* layer) {
```

Comment:
Taking `const Layer&` is probably a little safer than `Layer* layer` so you won't run into `nullptr` issues.

Reply:
Done in the latest commit. I also added a few more const tags to accessors and other inlines (at least one of which was needed to allow the layer parameter here to be marked const).